packages/server/modules/shared/authz.ts

```diff
@@ -0,0 +1,293 @@
+      context.scopes.indexOf('*') === -1
+    )
+      return authFailed(context, new FE('You do not have the required privileges.'))
+    return authSuccess(context)
+  }
+
+type StreamGetter = ({
+  streamId,
+  userId
+}: {
+  streamId: string
+  userId?: string
+}) => Promise<Stream>
+
+// this doesn't do any checks  on the scopes, its sole responsibility is to add the
+// stream object to the pipeline context
+export const contextRequiresStream =
+  (streamGetter: StreamGetter): AuthPipelineFunction =>
+  // stream getter is an async func over { streamId, userId } returning a stream object
+  // IoC baby...
+  async ({ context, authResult, params }) => {
+    if (!params?.streamId)
+      return authFailed(context, new CE("The context doesn't have a streamId"))
+    // because we're assigning to the context, it would raise if it would be null
+    // its probably?? safer than returning a new context
+    if (!context) return authFailed(context, new CE('The context is not defined'))
+
+    // cause stream getter could throw, its not a safe function if we want to
+    // keep the pipeline rolling
+    try {
+      const stream = await streamGetter({
+        streamId: params.streamId,
+        userId: context?.userId
+      })
+      if (!stream)
+        return authFailed(context, new BRE('Stream inputs are malformed'), true)
+      context.stream = stream
+      return { context, authResult }
+    } catch (err) {
+      // this prob needs some more detailing to not leak internal errors
+      const error = err as Error
+      return authFailed(context, new CE(error.message))
+    }
+  }
+
+export const allowForRegisteredUsersOnPublicStreamsEvenWithoutRole: AuthPipelineFunction =
+  async ({ context, authResult }) =>
+    context.auth && context.stream?.isPublic
+      ? authSuccess(context)
+      : { context, authResult }
+
+export const allowForAllRegisteredUsersOnPublicStreamsWithPublicComments: AuthPipelineFunction =
+  async ({ context, authResult }) =>
+    context.auth && context.stream?.isPublic && context.stream?.allowPublicComments
+      ? authSuccess(context)
+      : { context, authResult }
+
+export const allowAnonymousUsersOnPublicStreams: AuthPipelineFunction = async ({
+  context,
+  authResult
+}) => (context.stream?.isPublic ? authSuccess(context) : { context, authResult })
+
+export const authPipelineCreator = (
+  steps: AuthPipelineFunction[]
+): AuthPipelineFunction => {
+  const pipeline: AuthPipelineFunction = async ({
+    context,
+    params,
+    authResult = { authorized: false }
+  }) => {
+    for (const step of steps) {
+      ;({ context, authResult } = await step({ context, authResult, params }))
+      if (authHasFailed(authResult) && authResult?.fatal) break
+    }
+    // validate auth result a bit...
+    if (authResult.authorized && authHasFailed(authResult))
+      throw new Error('Auth failure')
+    return { context, authResult }
+  }
+  return pipeline
+}
+
+import Express from 'express'
```

Comment:
In ESM all imports should always be at the top, you can only invoke an import at other arbitrary locations if they're dynamic/async imports. During compilation I bet the import gets moved up there anyway, but to keep the source code more readable it would be nice to move it up there as well.